packages/library-base/src/commonMain/kotlin/io/realm/internal/BaseResults.kt

```diff
@@ -0,0 +1,168 @@
+            )
+        } catch (exception: RealmCoreException) {
+            throw genericRealmCoreExceptionHandler("Invalid syntax for query `$query`", exception)
+        }
+    }
+
+    override fun asFlow(): Flow<BaseResults<E>> {
+        realm.checkClosed()
+        return realm.owner.registerObserver(this)
+    }
+
+    override fun delete() {
+        // TODO OPTIMIZE Are there more efficient ways to do this? realm_query_delete_all is not
+        //  available in C-API yet, but should probably await final query design
+        //  https://github.com/realm/realm-kotlin/issues/84
+        RealmInterop.realm_results_delete_all(nativePointer)
+    }
+
+    /**
+     * Returns a frozen copy of this query result. If it is already frozen, the same instance
+     * is returned.
+     */
+    override fun freeze(frozenRealm: RealmReference): BaseResults<E> {
+        val frozenDbPointer = frozenRealm.dbPointer
+        val frozenResults = RealmInterop.realm_results_resolve_in(nativePointer, frozenDbPointer)
+        return instantiateResults(frozenRealm, frozenResults, clazz, mediator)
+    }
+
+    /**
+     * Thaw the frozen query result, turning it back into a live, thread-confined RealmResults.
+     */
+    override fun thaw(liveRealm: RealmReference): BaseResults<E> {
+        val liveDbPointer = liveRealm.dbPointer
+        val liveResultPtr = RealmInterop.realm_results_resolve_in(nativePointer, liveDbPointer)
+        return instantiateResults(liveRealm, liveResultPtr, clazz, mediator)
+    }
+
+    override fun registerForNotification(callback: io.realm.internal.interop.Callback): NativePointer {
+        return RealmInterop.realm_results_add_notification_callback(nativePointer, callback)
+    }
+
+    override fun emitFrozenUpdate(
+        frozenRealm: RealmReference,
+        change: NativePointer,
+        channel: SendChannel<BaseResults<E>>
+    ): ChannelResult<Unit>? {
+        val frozenResult = freeze(frozenRealm)
+        return channel.trySend(frozenResult)
+    }
+
+    override fun realmState(): RealmState = realm
+}
+
+/**
+ * TODO : query
+ */
+internal class ElementResults<E : RealmObject> constructor(
+    realm: RealmReference,
+    nativePointer: NativePointer,
+    clazz: KClass<E>,
+    schema: Mediator
+) : BaseResults<E>(realm, nativePointer, clazz, schema) {
+
+    override fun get(index: Int): E {
+        val link = RealmInterop.realm_results_get<Link>(nativePointer, index.toLong())
+        val model = mediator.createInstanceOf(clazz)
+        model.link(realm, mediator, clazz, link)
+        @Suppress("UNCHECKED_CAST")
+        return model as E
+    }
+
+    override fun instantiateResults(
+        realmReference: RealmReference,
+        nativePointer: NativePointer,
+        clazz: KClass<E>,
+        mediator: Mediator
+    ): BaseResults<E> = ElementResults(realm, nativePointer, clazz, mediator)
+}
+
+/**
+ * TODO : query
+ */
+internal class ScalarResults<E : Any> constructor(
```

Comment:
This looks a bit odd? Why do we need a ScalarResults.  It either a Query object or a simple value?

Reply:
I think we do. Before introducing this API we only supported `RealmResults<RealmObject>` and its `get` function would assume by default the value returned by `realm_results_get` was a `Link`. Now it can be either a `Link` or a primitive/scalar due to the addition of flows for aggregation queries. The only way we could make it with only one `RealmResults` implementation is if `RealmInterop.realm_results_get` were able to instantiate a `Link` from the interop and preserve the generic type on the library side, which unfortunate cannot be done due to the way things are structured across modules :/